src/ray/gcs/redis_context-inl.h

```diff
@@ -0,0 +1,159 @@
+  return std::make_pair(ip_port[0], std::stoi(ip_port[1]));
+}
+
+template <typename RedisContextType, typename RedisConnectFunctionType>
+std::pair<Status, std::unique_ptr<RedisContextType, RedisContextDeleter>>
+ConnectWithoutRetries(const std::string &address,
+                      int port,
+                      const RedisConnectFunctionType &connect_function) {
+  // This currently returns the errorMessage in two different ways,
+  // as an output parameter and in the Status::RedisError,
+  // because we're not sure whether we'll want to change what this returns.
+  RedisContextType *newContext = connect_function(address.c_str(), port);
+  if (newContext == nullptr || (newContext)->err) {
+    std::ostringstream oss;
+    if (newContext == nullptr) {
+      oss << "Could not allocate Redis context.";
+    } else if (newContext->err) {
+      oss << "Could not establish connection to Redis " << address << ":" << port
+          << " (context.err = " << newContext->err << ").";
+    }
+    return std::make_pair(Status::RedisError(oss.str()), nullptr);
+  }
+  return std::make_pair(Status::OK(),
+                        std::unique_ptr<RedisContextType, RedisContextDeleter>(
+                            newContext, RedisContextDeleter()));
+}
+
+template <typename RedisContextType, typename RedisConnectFunctionType>
+std::pair<Status, std::unique_ptr<RedisContextType, RedisContextDeleter>>
+ConnectWithRetries(const std::string &address,
+                   int port,
+                   const RedisConnectFunctionType &connect_function) {
+  RAY_LOG(INFO) << "Attempting to connect to address " << address << ":" << port << ".";
+  int connection_attempts = 0;
+  auto resp = ConnectWithoutRetries<RedisContextType>(address, port, connect_function);
+  auto status = resp.first;
+  while (!status.ok()) {
+    if (connection_attempts >= RayConfig::instance().redis_db_connect_retries()) {
+      RAY_LOG(FATAL) << RayConfig::instance().redis_db_connect_retries() << " attempts "
+                     << "to connect have all failed. Please check whether the"
+                     << " redis storage is alive or not. The last error message was: "
+                     << status.ToString();
+      break;
+    }
+    RAY_LOG_EVERY_MS(ERROR, 1000)
+        << "Failed to connect to Redis due to: " << status.ToString()
+        << ". Will retry in "
+        << RayConfig::instance().redis_db_connect_wait_milliseconds() << "ms.";
+
+    // Sleep for a little.
+    std::this_thread::sleep_for(std::chrono::milliseconds(
+        RayConfig::instance().redis_db_connect_wait_milliseconds()));
+    resp = ConnectWithoutRetries<RedisContextType>(address, port, connect_function);
+    status = resp.first;
+    connection_attempts += 1;
+  }
+  return resp;
+}
+
+RedisAsyncContext *CreateAsyncContext(
+    std::unique_ptr<redisAsyncContext, RedisContextDeleter> context) {
+  return new RedisAsyncContext(std::move(context));
+}
+
+template <typename ConnectType>
```

Comment:
This is RedisContextType?

Reply:
This is either `redisAsyncContext` or `MockRedisAsyncContext` and the reason it's even templated at all is so that we can control what `ConnectWithRetries` returns inside the `RedisResponseFn` inside the unit test (by defining a `ConnectWithRetries<MockRedisAsyncContext>`). 

It's also really bad because MockRedisAsyncContext is analogous to RedisAsyncContext (wrapper around the redisAsyncContext C struct) so it's one wrapper level above the other one which makes some of the mappings awkward in the unit test and `RedisResponseFn`

Reply:
Ah, you meant that it should be renamed to `RedisContextType`?

It's true that `RedisContextType` will always be the same as `ConnectType` when called from here, but in general `RedisContextType` can also be `redisContext` whereas `ConnectType` is always `redisAsyncContext` except in the case of unit testing.

Reply:
Yea I was thinking renaming to `RedisContextType`?

Reply:
Renamed! I still think they are different though